File: apps/api/src/alphadesk_api/schemas/accounting.py
"""Pydantic contracts for the M04 simulated-account APIs."""

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from alphadesk_domain.enums import (
    AccountStatus,
    AccountValuationStatus,
    CashLedgerEntryType,
    LedgerTransactionStatus,
    LedgerTransactionType,
    PositionLedgerEntryType,
    ReconciliationStatus,
    SettlementPolicy,
)


class DomainResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)


class AccountCreateRequest(BaseModel):
    account_code: str = Field(min_length=1, max_length=64)
    name: str = Field(min_length=1, max_length=128)
    base_currency: str = Field(default="CNY", min_length=1, max_length=8)
    initial_cash: Decimal = Field(default=Decimal("0"), ge=0)
    settlement_policy: SettlementPolicy = SettlementPolicy.IMMEDIATE
    idempotency_key: str = Field(min_length=1, max_length=128)


class AccountUpdateRequest(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=128)
    status: AccountStatus | None = None
    settlement_policy: SettlementPolicy | None = None


class FundingRequest(BaseModel):
    amount: Decimal = Field(gt=0)
    idempotency_key: str = Field(min_length=1, max_length=128)
    description: str | None = Field(default=None, max_length=500)


class AccountResponse(DomainResponse):
    id: UUID
    account_code: str
    name: str
    status: AccountStatus
    broker_type: str
    base_currency: str
    settlement_policy: SettlementPolicy
    created_at: datetime
    updated_at: datetime


class CashBalanceResponse(DomainResponse):
    id: UUID
    account_id: UUID
    currency: str
    total_cash: Decimal
    available_cash: Decimal
    frozen_cash: Decimal
    row_version: int
    as_of: datetime


class PositionResponse(DomainResponse):
    id: UUID
    instrument_id: UUID
    total_quantity: Decimal
    available_quantity: Decimal
    frozen_quantity: Decimal
    unsettled_quantity: Decimal
    cost_basis: Decimal
    average_cost: Decimal
    market_value: Decimal | None
    realized_pnl: Decimal
    unrealized_pnl: Decimal | None
    last_price: Decimal | None
    last_price_at: datetime | None
    valuation_status: AccountValuationStatus
    as_of: datetime


class AccountDetailResponse(AccountResponse):
    cash_balances: list[CashBalanceResponse]
    positions: list[PositionResponse]


class LedgerTransactionResponse(DomainResponse):
    id: UUID
    account_id: UUID
    transaction_type: LedgerTransactionType
    status: LedgerTransactionStatus
    business_key: str
    related_order_id: UUID | None
    related_fill_id: UUID | None
    correlation_id: UUID
    occurred_at: datetime
    posted_at: datetime | None
    description: str | None
    metadata: dict[str, Any]


class CashLedgerEntryResponse(DomainResponse):
    id: int
    ledger_transaction_id: UUID
    currency: str
    entry_type: CashLedgerEntryType
    total_delta: Decimal
    available_delta: Decimal
    frozen_delta: Decimal
    gross_amount: Decimal | None
    fee_amount: Decimal | None
    total_cash_after: Decimal
    available_cash_after: Decimal
    frozen_cash_after: Decimal
    related_fill_id: UUID | None
    occurred_at: datetime


class PositionLedgerEntryResponse(DomainResponse):
    id: int
    ledger_transaction_id: UUID
    instrument_id: UUID
    entry_type: PositionLedgerEntryType
    quantity_delta: Decimal
    cost_basis_delta: Decimal
    realized_pnl_delta: Decimal
    total_quantity_after: Decimal
    cost_basis_after: Decimal
    average_cost_after: Decimal
    realized_pnl_after: Decimal
    related_fill_id: UUID | None
    occurred_at: datetime


class AccountSnapshotResponse(DomainResponse):
    id: UUID
    account_id: UUID
    as_of: datetime
    cash_total: Decimal
    cash_available: Decimal
    cash_frozen: Decimal
    positions_cost_basis: Decimal
    positions_market_value: Decimal | None
    total_equity: Decimal | None
    realized_pnl: Decimal
    unrealized_pnl: Decimal | None
    valuation_status: AccountValuationStatus
    priced_position_count: int
    unpriced_position_count: int
    latest_price_time: datetime | None
    metadata: dict[str, Any]


class ReconciliationResponse(DomainResponse):
    id: UUID
    account_id: UUID
    status: ReconciliationStatus
    started_at: datetime
    completed_at: datetime | None
    expected_cash: dict[str, Any]
    actual_cash: dict[str, Any]
    expected_positions: dict[str, Any]
    actual_positions: dict[str, Any]
    discrepancy_count: int
    discrepancies: list[dict[str, Any]]


class PageResponse(BaseModel):
    items: list[Any]
    page: int
    page_size: int
    total: int


class AccountSummaryResponse(BaseModel):
    account: AccountResponse
    cash_balances: list[CashBalanceResponse]
    positions: list[PositionResponse]
    latest_snapshot: AccountSnapshotResponse | None
    latest_reconciliation: ReconciliationResponse | None

File: apps/web/src/types/accounting.ts
export type AccountStatus = "ACTIVE" | "READ_ONLY" | "SUSPENDED" | "CLOSED";
export type ValuationStatus = "COMPLETE" | "PARTIAL" | "STALE" | "UNAVAILABLE";

export interface Page<T> {
  items: T[];
  page: number;
  page_size: number;
  total: number;
}

export interface Account {
  id: string;
  account_code: string;
  name: string;
  status: AccountStatus;
  broker_type: string;
  base_currency: string;
  settlement_policy: "IMMEDIATE" | "T_PLUS_ONE";
  created_at: string;
  updated_at: string;
}

export interface CashBalance {
  id: string;
  account_id: string;
  currency: string;
  total_cash: string;
  available_cash: string;
  frozen_cash: string;
  row_version: number;
  as_of: string;
}

export interface Position {
  id: string;
  instrument_id: string;
  total_quantity: string;
  available_quantity: string;
  frozen_quantity: string;
  unsettled_quantity: string;
  cost_basis: string;
  average_cost: string;
  market_value: string | null;
  realized_pnl: string;
  unrealized_pnl: string | null;
  last_price: string | null;
  last_price_at: string | null;
  valuation_status: ValuationStatus;
  as_of: string;
}

export interface AccountSnapshot {
  id: string;
  account_id: string;
  as_of: string;
  cash_total: string;
  cash_available: string;
  cash_frozen: string;
  positions_cost_basis: string;
  positions_market_value: string | null;
  total_equity: string | null;
  realized_pnl: string;
  unrealized_pnl: string | null;
  valuation_status: ValuationStatus;
  priced_position_count: number;
  unpriced_position_count: number;
  latest_price_time: string | null;
  metadata: Record<string, unknown>;
}

export interface Reconciliation {
  id: string;
  account_id: string;
  status: "MATCHED" | "MISMATCHED" | "FAILED";
  started_at: string;
  completed_at: string | null;
  discrepancy_count: number;
  discrepancies: Record<string, unknown>[];
}

export interface AccountSummary {
  account: Account;
  cash_balances: CashBalance[];
  positions: Position[];
  latest_snapshot: AccountSnapshot | null;
  latest_reconciliation: Reconciliation | null;
}

export interface LivePositionValuation {
  instrument_id: string;
  symbol: string;
  quantity: string;
  price: string | null;
  market_value: string | null;
  unrealized_pnl: string | null;
  price_source: string | null;
  quote_time: string | null;
  freshness: "FRESH" | "STALE" | "MISSING";
}

export interface AccountLiveValuation {
  account_id: string;
  cash_balance: string;
  positions_market_value: string;
  total_equity: string;
  status: "COMPLETE" | "STALE" | "UNAVAILABLE";
  calculated_at: string;
  positions: LivePositionValuation[];
}

export interface LedgerTransaction {
  id: string;
  transaction_type: string;
  status: string;
  business_key: string;
  related_fill_id: string | null;
  occurred_at: string;
  description: string | null;
}

export interface CashLedgerEntry {
  id: number;
  entry_type: string;
  total_delta: string;
  total_cash_after: string;
  gross_amount: string | null;
  fee_amount: string | null;
  occurred_at: string;
}

export interface PositionLedgerEntry {
  id: number;
  instrument_id: string;
  entry_type: string;
  quantity_delta: string;
  cost_basis_delta: string;
  realized_pnl_delta: string;
  total_quantity_after: string;
  occurred_at: string;
}
